### PR TITLE
feat: new rules for @hasPermissions validation on default values 

### PR DESCRIPTION
**Pull request by Dorkt**
- Made @hasPermissions not check permissions on input values when they are equal to the default value;
- New tests for hasPermissions.test;

Closes: #71

### CODE REVIEW

File: lib/hasPermissions.ts [hunk 1]
```diff
@@ -203,16 +207,44 @@ export class HasPermissionsDirectiveVisitor<
...
+        value === (container as GraphQLArgument).defaultValue
+      ) {
+        return value;
+      }
+
       if (
         !permissions ||
         !permissions.length ||
-        (isUsedOnInputOrArgument && (value === null || value === undefined))
+        (isUsedOnInputOrArgument && value === undefined)
```
Comment (cabelitos) [inline lib/hasPermissions.ts:247]: why did you remove null from there?

File: lib/hasPermissions.ts [hunk 2]
```diff
@@ -203,16 +207,44 @@ export class HasPermissionsDirectiveVisitor<
...
+            DirectiveLocation.INPUT_OBJECT) &&
+        container instanceof GraphQLInputObjectType &&
+        path &&
+        value === container.getFields()[path[path.length - 1]].defaultValue
+      ) {
+        return value;
+      }
+
+      if (
+        location === DirectiveLocation.ARGUMENT_DEFINITION &&
```
Comment (cabelitos) [inline lib/hasPermissions.ts:229]: The conditions in here are the same from the if above. Just move this `location === DirectiveLocation.ARGUMENT_DEFINITION` to the `||` in the if above

File: lib/hasPermissions.ts [hunk 3]
```diff
@@ -203,16 +207,57 @@ export class HasPermissionsDirectiveVisitor<
...
       resolverSource: unknown,
       resolverArgs: Record<string, unknown>,
+      path?: Array<string>,
     ): unknown => {
+      if (
+        (location === DirectiveLocation.INPUT_FIELD_DEFINITION ||
+          location === DirectiveLocation.INPUT_OBJECT) &&
+        container instanceof GraphQLInputObjectType &&
+        path &&
+        value instanceof Array
```
Comment (cabelitos) [inline lib/hasPermissions.ts:222]: the correct way for checking for arrays is
`Array.isArray(value)`

File: lib/hasPermissions.ts [hunk 4]
```diff
@@ -203,16 +207,57 @@ export class HasPermissionsDirectiveVisitor<
...
+      ) {
+        const verify = [];
+        for (let i = 0; i < value.length; i += 1) {
+          if (
+            value[i] !==
+            container.getFields()[path[path.length - 1]].defaultValue[i]
```
Comment (cabelitos) [inline lib/hasPermissions.ts:228]: Please, do 
```
const lastField = container.getFields()[path[path.length - 1]];
for (...)
```

No need to access it on every for iteration.

On top of that, is it correct to access the `defaultValue[i]` ?

File: lib/hasPermissions.ts [hunk 5]
```diff
@@ -203,16 +207,57 @@ export class HasPermissionsDirectiveVisitor<
...
+          ) {
+            verify.push(false);
+          } else {
+            verify.push(true);
+          }
+        }
+        if (!verify.includes(false)) {
```
Comment (cabelitos) [inline lib/hasPermissions.ts:235]: this verify array looks strange. What are you trying to achieve? On top of that doing repeated searches on array (using the includes) is not very good. We should probably use a better approach.

File: lib/hasPermissions.ts (hunk 1 above)
Comment (Dorkt) [reply lib/hasPermissions.ts:260]: I'm sorry for just answer now, I didn't see this comment, but answering that, I removed this null, because now if the user doesn't have permissions to send the field, only the default value should be allowed (not even null should be allowed if null is not the default value), and with that `value === null`, the user can be pass `{ someFilter: null }` and receive some value from the query, even without permission.

File: lib/hasPermissions.ts (hunk 4 above)
Comment (Dorkt) [reply lib/hasPermissions.ts:228]: I think is correct to access the `defaultValue[i]`, because in some cases, the `defaultValue` is a Array, and is necessary check if all the values is equals, but thinking now, maybe is good to check if the `defaultValue` is a Array before, because if don't, no need to enter the loop.

And I changed the: 
`const lastField = container.getFields()[path[path.length - 1]];`
in advance, thanks for all the help

File: lib/hasPermissions.ts (hunk 5 above)
Comment (Dorkt) [reply lib/hasPermissions.ts:235]: I agree, and changed this approach to another, where I use just one variable to check this.

File: lib/hasPermissions.ts [hunk 6]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable no-console */
```
Comment (cabelitos) [inline lib/hasPermissions.ts:1]: please avoid such thing. It's a bad thing to disable eslint for no strong reason. let's not disable this rule.

File: lib/hasPermissions.ts [hunk 7]
```diff
@@ -203,16 +207,63 @@ export class HasPermissionsDirectiveVisitor<
...
+          }
+
+          if (verifyIfArraysAreEquals) {
+            return value;
+          }
+        }
+      }
+
+      if (
+        (location === DirectiveLocation.INPUT_FIELD_DEFINITION ||
```
Comment (cabelitos) [inline lib/hasPermissions.ts:245]: we should try to be repetitive.
We can merge this if and the above one like this.
```
const lastField = container.getFields()[path[path.length - 1]];

if (
  (location === DirectiveLocation.INPUT_FIELD_DEFINITION ||
    location === DirectiveLocation.INPUT_OBJECT) &&
  container instanceof GraphQLInputObjectType &&
  path &&
  path.length > 0 &&
  lastField
) {
  if (value === lastField.defaultValue) return value;
  if (Array.isArray(lastField.defaultValue) &&
      Array.isArray(value) &&
      value.length === lastField.defaultValue.length &&
      !value.every((item, idx) => item === lastField.defaultValue[idx])) {
      return value;
  }
}
```

Comment (Dorkt) [reply lib/hasPermissions.ts:245]: Okay, I just need before check if path exists and container is instanceof GraphQLInputObjectType, and I think maybe this exclamation is wrong in:

`!value.every((item, idx) => item === lastField.defaultValue[idx])) {`,

because if all data is the same, it returns true and should stop checking permissions.

File: lib/hasPermissions.ts [hunk 8]
```diff
@@ -200,20 +206,51 @@ export class HasPermissionsDirectiveVisitor<
...
+      path?: Array<string>,
+    ): unknown => {
+      if ('defaultValue' in container) {
+        return container.defaultValue;
+      }
+      /**
+       * Used this, because the container is verified before and just can be Argument or InputObjectType
+       * and the path in a InputObjectType just can be undefined when the input doesn't have variables inside
+       * but this is don't accepted for sintax GraphQL, so is very difficult test this, therefore, the last
+       * case of the function is very difficult to test too, when the return is undefined;
```
Comment (DDDKnightmare) [inline lib/hasPermissions.ts:220]: nit:
```suggestion
       * Used this because the container is verified before and just can only be `Argument` or `InputObjectType`.
       * The path in a `InputObjectType` can only be undefined when the input doesn't have any fields,
       * which is not accepted in GraphQL syntax.
       * due to the aforementioned reasons, it is difficult to test this case(if not impossible)
```

File: lib/hasPermissions.ts [hunk 9]
```diff
@@ -200,20 +206,51 @@ export class HasPermissionsDirectiveVisitor<
...
       resolverArgs: Record<string, unknown>,
+      path?: Array<string>,
     ): unknown => {
-      if (
-        !permissions ||
-        !permissions.length ||
-        (isUsedOnInputOrArgument && (value === null || value === undefined))
-      ) {
+      if (isUsedOnInputOrArgument) {
+        if (isEqual(value, undefined)) {
```
Comment (barbieri) [inline lib/hasPermissions.ts:243]: this is `value === undefined`

File: lib/hasPermissions.ts [hunk 10]
```diff
@@ -200,20 +206,51 @@ export class HasPermissionsDirectiveVisitor<
...
+       * case of the function is very difficult to test too, when the return is undefined;
+       */
+      /* istanbul ignore next */
+      if (isInputObjectType(container) && path && path.length > 0) {
+        const fieldName = path[path.length - 1];
+        const field = container.getFields()[fieldName];
+        return field.defaultValue;
+      }
+      /* istanbul ignore next */
+      return undefined;
```
Comment (DDDKnightmare) [inline lib/hasPermissions.ts:229]: can't this be reached when used on `ARGUMENT_DEFINITION` ?
e.g.:
```graphql
query {
  myQuery(myArgument: String @hasPermissions(permissions:["nope"]))
}
```

**Review by barbieri — DISMISSED** [review]

File: lib/hasPermissions.ts (hunk 10 above)
Comment (Dorkt) [reply lib/hasPermissions.ts]: No, because this example is reached with this part of the code: 

`if ('defaultValue' in container) {
    return container.defaultValue;
 }`

File: lib/hasPermissions.ts [hunk 11]
```diff
@@ -97,6 +103,21 @@ export type HasPermissionsDirectiveArgs = {
...
+): unknown => {
+  if ('defaultValue' in container) {
+    return container.defaultValue;
+  }
+  if (isInputObjectType(container) && path && path.length > 0) {
+    const fieldName = path[path.length - 1];
+    const field = container.getFields()[fieldName];
+    return field.defaultValue;
+  }
+  return undefined;
```
Comment (DDDKnightmare) [inline lib/hasPermissions.ts:118]: If no default value is not defined on the schema, it should be checked against null instead of undefined.

e.g.:
using the folowing schema
```graphql
input MyType {
  field: String @hasPermissions(permissions: ["x"])
}
query {
  myQuery(input: MyType!): Boolean
}
```
if the following input is used, the request should be accepted

```graphql
query Q1 {
  myQuery(input: {
    field: null
  })
}
```

```suggestion
  let defaultValue: unknown;
  if ('defaultValue' in container) {
    defaultValue = container.defaultValue;
  }
  if (isInputObjectType(container) && path && path.length > 0) {
    const fieldName = path[path.length - 1];
    const field = container.getFields()[fieldName];
    defaultValue = field.defaultValue;
  }
  return defaultValue ?? null;
```

Comment (DDDKnightmare) [reply lib/hasPermissions.ts]: I missed some points on the previous comment.
In some cases, `null` may not be accepted even if sent directly (e.g.: inputs used on mutations), whereas inputs used on queries can accept it without problems.
I think the best way to cover both cases would be adding a new argument to `@hasPermissions`
```graphql
input SomeMutationInput {
  fieldThatCannotReceiveNull: String @hasPermissions(permissions: ["x"])
  fieldThatCanReceiveNull: String @hasPermissions(permissions: ["x"], noDefaultEqualsNull: true)
}
input InputWhoseFieldsCannotReceiveNullWithoutPermission @hasPermissions(permissions: ["x"]){
  inputField: String
}
input SomeQueryInput @hasPermissions(permissions: ["y"], noDefaultEqualsNull: true) {
  fieldThatCanReceiveNull: String
}
mutation {
  updateSomething(input: SomeMutationInput!, otherInput: InputWhoseFieldsCannotReceiveNullWithoutPermission): String
}
query {
  someQuery(input: SomeQueryInput): Boolean
}
```
This way, one can choose which inputs allows explicitly sending null when defining the schema.
Using the example schema above and an user that doesn't have the necessary permissions to send the fields, the following should work:
```graphql
query myQuery {
  someQuery(input: {
    fieldThatCanReceiveNull: null
  })
}

mutation myMutation {
  updateSomething(input: {
    fieldThatCanReceiveNull: null
  })
}
```
And the following shouldn't
```graphql
mutation myMutation {
  updateSomething(input: {
    fieldThatCannotReceiveNull: null
  })

  updateSomething(otherInput: {
    inputField: null
  })
}
```

Note that for fields that have objects, checking this is more complex, but using `lodash.isEqualWith` may suffice.
e.g.:
```typescript
const compareWithDefaultAsNull = (value: unknown, default: unknown): boolean | undefined => {
  if(default === undefined){
    return value === null;
  }
  // delegate to underlying isEqual
  return undefined
}
const compareWithDefaultAsIs = (_value: unknown, _default: unknown): undefined => undefined

//...
let compareFn = compareWithDefaultAsIs;
if(noDefaultEqualsNull){
  compareFn = compareWithDefaultAsNull 
}
if(isEqualWith(value,  defaultValue, compareFn)){
  return value
}
// checkPermissions
```

Comment (barbieri) [reply lib/hasPermissions.ts]: @DDDKnightmare arguments shouldn't have any difference if they are used on mutations or queries, I'd avoid messing with that as you propose, instead just check for undefined and that's it. No extra arguments or the likes. The schema may declare `= null` as a default value if that's desired.

We need to reduce the variants and make it more consistent, your proposed change will make it less consistent and less predictable.

**Review by DDDKnightmare — APPROVED** [review]
